File: jclasslib/src/org/gjt/jclasslib/browser/detail/constants/AbstractConstantInfoDetailPane.java
/*
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
    License as published by the Free Software Foundation; either
    version 2 of the license, or (at your option) any later version.
*/

package org.gjt.jclasslib.browser.detail.constants;

import org.gjt.jclasslib.browser.BrowserServices;
import org.gjt.jclasslib.browser.detail.FixedListDetailPane;

import javax.swing.tree.TreePath;

/**
    Base class for all constant pool entry detail panes.
 
    @author <a href="mailto:jclasslib@ej-technologies.com">Ingo Kegel</a>
    @version $Revision: 1.5 $ $Date: 2003-08-18 08:16:21 $
*/
public abstract class AbstractConstantInfoDetailPane extends FixedListDetailPane {

    /** Message which is diplayed if the constant pool entry is invalid. */
    protected static final String MESSAGE_INVALID_CONSTANT_POOL_ENTRY = "invalid constant pool entry";

    /**
        Constructor.
        @param services the associated browser services.
     */
    protected AbstractConstantInfoDetailPane(BrowserServices services) {
        super(services);
    }
    
    /**
        Get the constant pool index corrensponding to a selection in 
        <tt>BrowserTreePane</tt>.
        @param treePath the tree path
        @return the index
     */
    protected int constantPoolIndex(TreePath treePath) {
        return getIndex(treePath);
    }
    
}

File: modules/browser/src/main/java/org/gjt/jclasslib/browser/detail/constants/AbstractConstantInfoDetailPane.kt
/*
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
    License as published by the Free Software Foundation; either
    version 2 of the license, or (at your option) any later version.
*/

package org.gjt.jclasslib.browser.detail.constants

import org.gjt.jclasslib.browser.BrowserServices
import org.gjt.jclasslib.browser.detail.FixedListDetailPane
import org.gjt.jclasslib.structures.Constant
import java.util.*

import javax.swing.tree.TreePath

abstract class AbstractConstantInfoDetailPane<T : Constant>(services: BrowserServices) : FixedListDetailPane(services) {

    private val showHandlers = ArrayList<(constant: T) -> Unit>()

    protected fun getConstant(treePath: TreePath): T {
        return constantClass.cast(getElement(treePath))
    }

    abstract val constantClass: Class<T>

    protected fun addConstantPoolLink(name: String, indexResolver: (constant: T) -> Int) {
        val nameLabel = linkLabel()
        val nameVerboseLabel = highlightLabel()
        addDetailPaneEntry(normalLabel(name), nameLabel, nameVerboseLabel)
        showHandlers.add { constant ->
            constantPoolHyperlink(nameLabel, nameVerboseLabel, indexResolver(constant))
        }
    }

    override fun show(treePath: TreePath) {
        super.show(treePath)
        val constant = getConstant(treePath)
        showHandlers.forEach { it.invoke(constant) }
    }
}


